### ISSUE TITLE
KDF135 X9.63 - issue with odd fieldSize

### ISSUE THREAD

**Issue by dghgit** [usnistgov/ACVP#886]
**environment**
Demo

**testSessionId**
89799

**vsId**
233377

**Algorithm registration**
{"algorithm":"kdf-components","mode":"ansix9.63","revision":"1.0","prereqVals":[{"algorithm":"SHA","valValue":"same"}],"sharedInfoLength":[0,1024],"fieldSize":[224,409],"keyDataLength":[128,1024],"hashAlg":["SHA2-224","SHA2-256","SHA2-384","SHA2-512"]}

**Endpoint in which the error is experienced**
https://demo.acvts.nist.gov:443
[23377.zip](https://github.com/usnistgov/ACVP/files/4761668/23377.zip)
/acvp/v1/testSessions/89799/vectorSets/233377/results GET

**Expected behavior**
Vectors for fieldSize 409, 521, 571 all fail using SHA2-512, I would have expected them to pass.

It's the first element of: tgId:14, described as:

{"tgId":14,"hashAlg":"SHA2-512","sharedInfoLength":0,"keyDataLength":128,"fieldSize":409,"testType":"AFT","tests":[{"tcId":261,"z":"0CCE1DBF51AC56133C5D4FACE7E0948A87C87872E10E3AF01EDA92094753D0B20CE210ABD4941DBA5D274B478D1930F6198B6880","sharedInfo":""}

the response my software is generating is:
{"tcId":261,"keyData":"297a62a750a2603be92695c4f2a08bfc"}

the sample data contains:
{"tcId":261,"keyData":"A6A5F044EE8991DF8DD8416125AF30ED"}

I have a feeling it's something about how I'm interpreting the fieldSize, it seems it's not equivalent the length parameter in the original ACVP test. The other thing that puzzles me though is that the bit length of the input is either 412 or 416 (big endian or little endian) so it's actually outside of the fieldsize. This does make me wonder if I should be truncating somehow but having tried a few guesses I have had to admit defeat.

Would someone please tell me how I should be interpreting the z value and the fieldsize in this context. Thanks.

**Additional context**
If I restrict my testing to fieldsizes which are multiples of 8, everything passes.

**Comment by celic** [comment]
Take the leftmost `fieldsize` bits of `z`. All hexadecimal for us is an even number of characters, so it will always represent a multiple of 8 bits. For values that are not a multiple of 8 bits, some truncation needs to happen.

**Comment by dghgit** [comment]
Thanks for the hint. I'm still not there yet. The X9.31 KDF requires z to be an octet string for input (well according to SEC-1 and ISO/IEC 18033-2:2006(E) at any rate, I'll admit I don't have the X9.31 standard handy, but SP 800-135 also refers to SEC-1), so in this case zero bit padding would need to be added somewhere to produce a string of 416 bits. I think the issue I'm having is with the padding.

Reading the hint I would interpret this as meaning I would "and" the last byte of z with 0x80, taking the left most bit as the start of the octet string -  which in this case would not change it's value. Would you tell me what I'm doing wrong - it still doesn't match. Thanks.

**Comment by celic** [comment]
Z is always has the length of exactly the fieldSize. So if you are using an odd-length field size, you must be able to handle incomplete bytes as the input to the hash function. In ANS x9.63 (which is what I think you are referring to) there is no reference to ensuring that Z is padded to a complete byte.

**Comment by dghgit** [comment]
Okay, that explains the issue, we're all octet aligned here.

I am curious about one thing though, SP 800-135 states "Note that any KDF that meets the specification of either the ANS X9.42-2001 or ANS X9.63-2001 hash-based KDF and is used in a scheme specified in one of these two Standards is approved when conditions (2) and (3) above are met. For example, the KDF specified in Section 2.1.2 (Generation of Keying Material) of RFC 2631 and the KDF (specified in [SEC1]) used in RFC 3278 are approved. " 

Section 3.1 of SEC-1 states "Keying data should be calculated using ANSI-X9.63-KDF as follows:... An octet string Z which is the shared secret value." RFC 3278, and more recently, RFC 8418 also refers to the algorithm described in SEC-1 as being equivalent. 

Further to this, here is an example from the CAVP:

[SHA-512]
[shared secret length = 571]
[SharedInfo length = 1024]
[key data length = 4096]

COUNT = 0
Z = 03331b883f01f41568a41b9314964e6fc8f11a076cc0ef79fd1cf55e3ade5fd85b1dc88c146d0350c7017d928baad06656ec4e2457e80d91ae6d7bc9b5d5583b80f5f5756b51d85c
SharedInfo = d4964cdb1b47549e3be0d495eff88469d6e187e3f5e1f53530ebcec125bb4df405d31057952362b16003caffde63c2563cfdcc40e905a57bc38fcec6ae5bf2f677532e8ece00cfb1e4b400cc8b8ae5dfeea0e8b21f21f9c8daa15a233b0466e75431608b311780f3472ce81b58c8f8d88711dd43cbdba43a6e337c5e5fdfb438

With the correct answer:

key_data = dcf73e3170e837422026eed28e2acc1cd5dfc5ef98142153006dafc3c78c6097d3e7934dabd1701129fe469f76f59ad43957182b1bfc464541384017fb7cd6609444b617851dba281820d4847cd44303752bfc25a395353fe478c39801ba4bc8f26e34db041d86ab5cd77dff91b8539eafa6938200d0410b011855e9530401810333808d5133217d01dc1fd7d1a0f744a2302160053c7e2f0ff6867112e9949f1e886691f6ede850134bceb9e3f6bc10c7ff68822f7bce16aa37c682b4fd4bfb1ac751a5a7eeb99c8a948335a2a8e61d4858c4974d85f422f68608713d1eab5d782248e5875d6dd2dc2176131cf5a4cd0cf7571b90740588f84ed9a6733cd10b8d57b41a6c4c912ce9c6dede9bf9e7e57ee91c5377066e543eacd6b8bb0a18ca77839ede352edd9668d49ece403098686487fd228576c2fd9e1584da626ed6a736760754b56af8e0ef65be215a336f1c5049668e08cf78253266d613ab354524abc8ac02f42885f3eba0691e92584deb030c5b7ec2ec3f0bb26a7485256b790623b6210267d709102b2d0a7b073d0750460a5efbbdcbc0072a9ac41eadd434d1e666e900da1e219510ba5833f07cafbf8b2adde39f08cf9b49134aa8f1b833d11dfa0dc8427857e5c8629615d364b449392aae59e2d82db69aa42ae2a65781533f748f55b807495387f65f8b240458d324f51e3b2cf9b729968faaad045fba56

This is from a test set dated in 2015. Under the new interpretation this vector and its accepted answer is now incorrect.

I guess my point is, that given long standing practice from the CAVP, together with what is stated in SP 800-135 the use of the octet string is already established as the FIPS compliant interpretation, so while ANSI x9.63 (sorry about the typo, standards overload...) may not be specific about the use of octets, that is what this algorithm means in the context of SP 800-135. 

I do understand that the ACVP server isn't strictly about validating FIPS algorithms, but could it be noted somewhere that odd fieldsizes in this test should not be used for the purposes of SP 800-135 certification testing as they deviate from what is the already established and documented practice for implementing the KDF, or better still would it be possible to add a conformance clause to the test, allowing people to test the algorithm as it has been validated through the CAVP for all field sizes?

I fear the current approach is going to cause a lot of confusion and compliance issues.

**Comment by celic** [comment]
Unfortunately ANS x9.63 doesn't contain any examples of the KDF being performed to observe Z. If CAVS in the past used a left-padding on Z, then ACVP should do the same. 

ANS x9.63 calls Z a bit string that is the shared secret value. Unfortunately it does not reference it as an octet, just a series of bits. At that point, it can be represented in any agreed upon fashion. ACVP uses the mentioned way of communicating bit strings with an incomplete amount of bytes. The standard itself seems to be ambiguous in this way. 

I'll discuss this a bit internally. Obviously we will continue to utilize this testing for SP800-135r1 validations. Nothing in a standard has been pointed out stating that either method of communicating Z is correct. 

Regardless, I'll see if I can update ours to match how CAVS performed these tests.

**Comment by celic** [comment]
I have updated ACVP to operate as CAVS did. Interestingly it seems like the ANS 9.63 KDF uses ALL bits in the shared secret Z even though the shared secret Z is supposed to be the same length as the field size. The leftmost padding bits end up getting hashed along with the shared secret Z in order to form the final result. I'm curious if this is proper according to the standard, but it is so ambiguous it is unclear. 

Long story short, in the next demo deployment you will see this behavior. Please test it when you receive notice from Jason Arnold that it was deployed.

**Comment by dghgit** [comment]
Will do. I appreciate the thoughtful consideration of this. Thanks.

**Comment by celic** [comment]
Fixed with patch today.
